The GC root placement pass 1.0 deserves

Keno opened the pull request:
Item 1/2 on my 1.0 must have list:

This is a principled rewrite of our GC root placement pass to occur late in the pass pipeline rather than early as well as produce a more optimal placement of GC roots. Additionally, it will allow us to perform much more general transformations at the LLVM stage, because we no longer have to worry about potentially introducing unrooted values, etc.  It also allows the LLVM optimizer itself to do significantly more aggressive optimizations in the presence of boxed objects.

Currently this is written again LLVM 5.0 and does use features that are new to that release, as well as the following LLVM patches:
https://reviews.llvm.org/D32593
https://reviews.llvm.org/D33110
https://reviews.llvm.org/D33129
https://reviews.llvm.org/D33179

I do think however that I can backport this to 4.0, albeit inserted closer the start of the pipeline. That wouldn't give quite as much benefit, but it should be equivalent or slightly better to what we have right now.

I have focused on correctness over compile time performance in the implementation here and I know the algorithm is not optimal (it's fair to say it's pretty naive). Additionally, there is a significant number of follow on optimizations, as well as codegen cleanup that is yet to be done. However, this is already a pretty big change, so I'm saving those for after this is merged.

Design notes are in the devdocs, implementation notes as code comments in the placement pass. There's also tests for all the tricky corner cases I found.

Fixes #20981
Fixes #15369
Supersedes #21158

Keno committed:
Introduce new GC root placement pass

Design notes are in the devdocs. Algorithmic documentation in code comments.

Keno committed:
Remove old GC placement pass

vtjnash commented on src/codegen.cpp:
```diff
@@ -2362,13 +2370,15 @@ static void simple_escape_analysis(jl_value_t *expr, bool esc, jl_codectx_t *ctx
 static Value *emit_local_root(jl_codectx_t *ctx, jl_varinfo_t *vi)
 {
-    CallInst *newroot = CallInst::Create(prepare_call(gcroot_func), "", /*InsertBefore*/ctx->ptlsStates);
+    Instruction *newroot = new AllocaInst(T_prjlvalue, 0, "gcroot", /*InsertBefore*/ctx->ptlsStates);
```
I think for performance we will want to do:
```c
#ifndef NDEBUG
newroot->setName("gcroot");
#endif
```

Keno replied:
Probably.

timholy commented:
Sounds really nice. Does this have any implications for temporary buffer reuse? (For example, if the root can be placed in the caller...) I'm thinking of functions along the lines of
```julia
function mymedian(v)
    vc = copy(v)
    sort!(vc)  # just for illustration, this is not optimal...
    inds = linearindices(vc)
    vc[(first(inds)+last(inds)) ÷ 2]
end
```
for which, if you call it repeatedly with many small vectors, currently exacts a non-trivial GC cost that can be mitigated by pre-allocating a temporary buffer and passing it in as an argument.

vtjnash commented on src/codegen.cpp:
```diff
@@ -2387,17 +2397,27 @@ static void mark_gc_use(const jl_cgval_t &v)
     int slot = 0;
     assert(args.size() > 0);
+    auto lifetime_start = Intrinsic::getDeclaration(jl_Module, Intrinsic::lifetime_start, {T_pprjlvalue});
+    builder.CreateCall(lifetime_start, {
```
LLVM37+ syntax here, but elsewhere you seem to be maintaining LLVM33 support. Seems like the PR should just be consistent about dropping support?

vtjnash commented on src/codegen.cpp:
```diff
@@ -2625,12 +2651,13 @@ static Value *emit_f_is(const jl_cgval_t &arg1, const jl_cgval_t &arg2, jl_codec
     JL_FEAT_REQUIRE(ctx, runtime);
-    Value *varg1 = boxed(arg1, ctx);
-    Value *varg2 = boxed(arg2, ctx, false); // potentially unrooted!
+    Value *varg1 = mark_callee_rooted(boxed(arg1, ctx));
```
this isn't callee-rooted (and the creation of `varg2` could allocate)

Keno replied:
That's fine, the pass will still keep it live at any safepoint that it's life *across*. It just won't put it in a gc root if a safepoint terminates the live interval.

vtjnash commented on src/codegen.cpp:
```diff
@@ -2693,12 +2720,13 @@ static bool emit_builtin_call(jl_cgval_t *ret, jl_value_t *f, jl_value_t **args,
         if (jl_subtype(ty, (jl_value_t*)jl_type_type)) {
             *ret = emit_expr(args[1], ctx);
-            Value *rt_ty = boxed(emit_expr(args[2], ctx), ctx);
+            Value *rt_ty = mark_callee_rooted(boxed(emit_expr(args[2], ctx), ctx));
```
This isn't callee-rooted unless `boxed` tells you that it was

Keno replied:
Same comment as above. The placement pass figures it out. The callee rooted annotation is just to annotate that the only thing you're allowed to do with it is pass it to a function that callee roots.

vtjnash replied:
OK, but type-assert also doesn't root this, so it does need to be rooted at the call safepoint by the caller / codegen

vtjnash commented on src/codegen.cpp:
```diff
@@ -3351,22 +3386,26 @@ static jl_cgval_t emit_call_function_object(jl_method_instance_t *li, const jl_c
-                        retval.V),
-                    retval.gcroot);
+                builder.CreateStore(box,
+                                      retval.gcroot);
```
alignment here seems odd. seems like this could fit on the one line

vtjnash commented on src/jitlayers.cpp:
```diff
@@ -129,12 +130,16 @@ void addOptimizationPasses(PassManager *PM)
         PM->add(createAlwaysInlinerLegacyPass()); // Respect always_inline
 #else
         PM->add(createAlwaysInlinerPass()); // Respect always_inline
 #endif
```
are we actually confident that this can't break your gc-invariants and end up trying to run that pass on external code that it wasn't supposed to try to run on?

Keno replied:
Yes. It only looks at our address spaces. If external code uses our address spaces, it needs to follow our invariants.

vtjnash commented on src/jitlayers.cpp:
```diff
@@ -489,8 +501,13 @@ JuliaOJIT::JuliaOJIT(TargetMachine &TM)
         addOptimizationPasses(&PM);
     }
     else {
```
could we write this as `addOptimizationPasses(&PM, /* -O */0)`, and make the `jl_options.opt_level` option to that function explicit?

Keno replied:
Yes.

vtjnash commented on doc/src/devdocs/llvm.md:
````diff
@@ -75,3 +75,159 @@ study it and the pass of interest in isolation.
+```
+The GC root placement pass will treat the jl_roots operand bundle as if it were
+a regular operand. However, as a final step, after the gc roots are inserted,
+it will drop the operand bundle to avoid confusing codegen.
````
"avoid confusing machine instruction selection"?

Keno commented:
> Sounds really nice. Does this have any implications for temporary buffer reuse?

Yes and no. It makes it a lot easier to do that kind of transformation at the LLVM level, as well as making LLVM-level inlining easier. Those two combined could easily achieve what you want in your example. However, by itself this doesn't really do much there.

StefanKarpinski commented on doc/src/devdocs/llvm.md:
```diff
@@ -75,3 +75,159 @@ study it and the pass of interest in isolation.
+
+Minimize the number of needed gc roots/stores to them subject to the constraint
+that at every safepoint, any live gc-tracked pointer (i.e. is a path after this
+point that contains a use of this pointer).
```
Incomplete sentence?

Keno replied:
yes, should have `is in a gc slot` ;).

StefanKarpinski commented on doc/src/devdocs/llvm.md:
```diff
@@ -75,3 +75,159 @@ study it and the pass of interest in isolation.
+for the function. As a result, the external rooting must be arranged while the
+value is still tracked by the system. I.e. it is not valid to attempt use the
+result of this operation to establish a global root - the optimizer may have
+already dropped the value.
```
Would it make any sense for `pointer_from_objref` to return a pointer in a GC-aware address space? Would that allow LLVM to automatically make code using such pointers safe(r)?

Keno replied:
No, the whole point of that function is to escape the GC.

yuyichao replied:
This escape semantic would also be required so that we can stack allocate `RefValue{bitstype}` in ccall.

StefanKarpinski commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1343 @@
+
+   Minimize the number of needed gc roots/stores to them subject to the constraint
+   that at every safepoint, any live gc-tracked pointer (i.e. is a path after this
+   point that contains a use of this pointer).
```
Same incomplete sentence.

StefanKarpinski commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1343 @@
+      This step performs necessary cleanup before passing the IR to codegen. In
+      particular, it removes any calls to julia_from_objref intrinsics and
+      removes the extra operand bundles from ccalls. In the future it could
+      also strip the addressspace information from all values as this
```
I don't think "address space" is one word

Keno replied:
Well the llvm annotation is `addrspace`, so I should probably just write that.

StefanKarpinski commented:
Great documentation and comments, btw. This is the standard we should all strive for :)

Keno commented:
Rebased and backported to 3.9.1. On 3.9.1, it needs to run very early in the pipeline, so the improvements are not nearly as large. So I guess just think that that as an incentive to move to 5.0 post haste. I don't intent to support any prior LLVM version.

StefanKarpinski commented:
We should definitely move master to LLVM 5 ASAP since we know that causes all sorts of havoc and needs a long time to settle into stability.

timholy commented:
It's also a good thing to do this before we introduce too many other incompatibilities: the package ecosystem is a good test bed for shaking out LLVM bugs, but most people won't update packages to 0.7 until 0.7 is looking imminent. Consequently, the longer we wait, the more packages will be untestable for reasons unrelated to the LLVM update.

Keno commented:
Fixed the 32bit build. The windows build seems to be failing because the LLVM patches I added here aren't being applied. @tkelman how do we rebuild the windows CI LLVM binaries?

tkelman commented:
I have to build them by hand, though now that we build llvm as a shared library we could maybe get them from the nightlies (with some fiddling needed for headers?)

Keno commented:
Not sure what's up with the OS X build. Passes locally. I suspect it'll also need an LLVM version bump.

tkelman commented:
OS X travis gets its llvm binary from the juliadeps homebrew tap

Keno commented:
@tkelman @staticfloat the patches here have been merged on master. Can we build new copies of LLVM for Win/OSX from the patchset that's currently on master?

staticfloat commented:
@keno can you open a PR to include the patches in [this file](https://github.com/staticfloat/homebrew-julia/blob/master/llvm39-julia.rb)?

Keno commented:
@staticfloat let me know when the bottles are built so I can kick off CI again here.

staticfloat commented:
Hmmm.  Compilation failed:

```
/tmp/llvm39-julia-20170525-79976-1eabd2f/llvm-3.9.1.src/tools/lld/include/lld/Core/Parallel.h:125:14: error: no member named 'thread' in namespace 'std'
        std::thread([=] {
        ~~~~~^
/tmp/llvm39-julia-20170525-79976-1eabd2f/llvm-3.9.1.src/tools/lld/include/lld/Core/Parallel.h:123:10: error: no member named 'thread' in namespace 'std'
    std::thread([&, threadCount] {
```

Is that coming from one of our patches?  This is being built with `Clang: 7.0 build 700`.

Keno commented:
`llvm-3.9.0_threads` perhaps?

staticfloat commented:
Ah, I see what's going on.  We have further patches we need to include to get rid of the `std::thread` stuff within `lld` because the llvm3.9 formula installs it by default.

Is there any reason I should get `lld` compiling properly, or should I just remove it from the subprojects that get built?

tkelman commented:
Do we need to be building lld currently? It may prove useful at some point but we haven't put anything for it in the makefiles here yet

staticfloat commented:
I just removed `lld` and `clang`, `polly`, and the other stuff we don't really need.  The new bottles are uploaded, so go for it.

Keno commented:
Rebased. @tkelman what's the status of the windows LLVM build?

vtjnash commented on src/cgutils.cpp:
```diff
@@ -1006,20 +1067,21 @@ static std::pair<Value*, bool> emit_isa(const jl_cgval_t &x, jl_value_t *type, c
     // intersection with Type needs to be handled specially
     if (jl_has_intersect_type_not_kind(type)) {
-        Value *vx = boxed(x, ctx);
+        Value *vx = mark_callee_rooted(boxed(x, ctx));
```
this isn't callee rooted (except for the case where it gets passed to `jltypeassert_func`)

Keno replied:
I included not-a-safepoint in the definition of callee rooted for now because it's the same for all practical purposes. In a follow on commit, I plan to make that notion explicit. At that point, I'll clean this up.

vtjnash commented on src/cgutils.cpp:
```diff
@@ -1049,15 +1112,20 @@ static std::pair<Value*, bool> emit_isa(const jl_cgval_t &x, jl_value_t *type, c
 #if JL_LLVM_VERSION >= 30700
-            builder.CreateCall(prepare_call(jlsubtype_func), { vxt, literal_pointer_val(type) }),
+            builder.CreateCall(prepare_call(jlsubtype_func),
+              { mark_callee_rooted(vxt),
```
jl_subtype doesn't root its argument

Keno replied:
Same reason as above.

vtjnash commented on src/cgutils.cpp:
```diff
@@ -2017,7 +2102,7 @@ static Value *box_union(const jl_cgval_t &vinfo, jl_codectx_t *ctx, const SmallB
             mark_gc_use(vinfo);
         }
-        box_merge->addIncoming(emit_bitcast(vinfo.V, T_pjlvalue), defaultBB);
+        box_merge->addIncoming(builder.CreateLoad(vinfo.gcroot), defaultBB);
```
should be `.V` still?

Keno replied:
Since you need to end up with a boxed value, I don't this they would ever disagree here.

vtjnash commented on src/cgutils.cpp:
```diff
@@ -2032,12 +2117,12 @@ static Value *boxed(const jl_cgval_t &vinfo, jl_codectx_t *ctx, bool gcrooted)
     if (vinfo.isboxed) {
         assert(vinfo.V && "Missing value for box.");
-        return vinfo.V;
+        return vinfo.gcroot ? builder.CreateLoad(vinfo.gcroot) : vinfo.V;
```
same here?

vtjnash commented on src/codegen.cpp:
```diff
@@ -2362,13 +2370,15 @@ static void simple_escape_analysis(jl_value_t *expr, bool esc, jl_codectx_t *ctx
 {
-    CallInst *newroot = CallInst::Create(prepare_call(gcroot_func), "", /*InsertBefore*/ctx->ptlsStates);
+    Instruction *newroot = new AllocaInst(T_prjlvalue, 0, "gcroot", /*InsertBefore*/ctx->ptlsStates);
+    // CallInst *newroot = CallInst::Create(prepare_call(gcroot_func), "", /*InsertBefore*/ctx->ptlsStates);
```
reminder to delete this dead code

vtjnash commented on src/codegen.cpp:
```diff
@@ -2387,22 +2397,51 @@ static void mark_gc_use(const jl_cgval_t &v)
+    // CallInst *largs = CallInst::Create(prepare_call(jlcall_frame_func),
+    //         ConstantInt::get(T_int32, args.size()),
+    //         "",
+    //         /*InsertBefore*/ctx->ptlsStates);
```
reminder to delete

Keno commented:
Found the OS X issue. @tkelman can you get the updated LLVM for windows in place?

Keno commented on doc/src/devdocs/llvm.md:
```diff
@@ -75,3 +75,159 @@ study it and the pass of interest in isolation.
+  pass will look for these to identify the base pointer for any use.
+- 10->12: Addressspace 12 servers merely as a hint that a GC root is not
+  required. However, do note that the 11->12 decay is prohibited, since
+  pointers should generally be storable to a GC slot, even in this address space.
```
Well, they are named in the code. I guess I should use those names here

vtjnash replied:
nice, this reads better without the numbers too

Keno commented on doc/src/devdocs/llvm.md:
```diff
@@ -75,3 +75,159 @@ study it and the pass of interest in isolation.
+
+Now let us consider what constitutes a use:
+- Loads whose loaded values is in one of the address spaces
+- Stores of a value in one of the address spaces to a location
```
yes, will add a separate bullet point

Keno commented on src/ccall.cpp:
```diff
@@ -1646,8 +1654,16 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
         JL_GC_POP();
-        return mark_or_box_ccall_result(emit_bitcast(ary, lrt),
-                                        retboxed, rt, unionall, static_rt, ctx);
+        if (lrt != T_prjlvalue) {
```
Fair enough.

vtjnash commented on src/codegen.cpp:
```diff
@@ -3781,14 +3837,19 @@ static void emit_assignment(jl_value_t *l, jl_value_t *r, jl_codectx_t *ctx)
+                    if (slot.gcroot) {
                         gcroot = emit_local_root(ctx);
-                    else
+                        box = builder.CreateLoad(slot.gcroot);
```
`load(gcroot) != boxed(slot)` in general. in particular, slot could represent a gep (getfield) or bitcast.

vtjnash commented on src/codegen.cpp:
```diff
@@ -4721,10 +4789,12 @@ static Function *gen_cfun_wrapper(jl_function_t *ff, jl_value_t *jlrettype, jl_t
+#if JL_LLVM_VERSION >= 50000
+            0,
+#endif
+            ConstantInt::get(T_int32, nargs +1), "jlcall", ctx2.ptlsStates);
```
missing space after `+`

vtjnash commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+    int MaxAssignedColor = -1;
+    while ((ActiveElement = Ordering.next()) != -1) {
+        assert(Colors[ActiveElement] == -1);
+        BitVector UsedColors(MaxAssignedColor + 2, false);
```
I think it might be good to move this malloc outside the while loop and just reset its size here (`UsedColors.reset().resize(MaxAssignedColor + 2)`)

Keno commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+   1. Local Scan
+
+      During this step, each Basic Block is inspected and analyzed for local
+      properties. In particular, we want to determine:
```
yes

Keno commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+{
+    if (!Phi->getType()->isPointerTy())
+        return;
+    Type *ElType = cast<PointerType>(Phi->getType())->getElementType();
```
Well, it could check the store type for every phi, or it could do something better. I think it's fine for now.

Keno commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+                NoteOperandUses(S, BBS, I, BBS.UpExposedUsesUnrooted);
+            } else if (auto *ASCI = dyn_cast<AddrSpaceCastInst>(&I)) {
+                if (getValueAddrSpace(ASCI) == AddressSpace::Tracked)
+                    MaybeNoteDef(S, BBS, ASCI, BBS.Safepoints);
```
It's not a safepoint, but I do treat it as a potential def here. In the future, I want to be a be more aggressive and not number these, but for now this is safer.

Keno commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+        for (int Live : S.LiveIfLiveOut[idx]) {
+            if (HasBitSet(BBS.LiveOut, Live))
+                S.LiveSets[idx][Live] = 1;
+        }
```
`LiveIfLiveOut` is not a BitVector

Keno commented:
I've addressed all the comments, except for some of the potential efficiency improvements and changing where the rooting occurs. I think those can be follow on changes.

vtjnash commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+                    }
+                    if (!V->getType()->isPointerTy())
+                        continue;
+                    if (isSpecialPtr(cast<PointerType>(V->getType())->getElementType())) {
```
also, since it's assumed below, might be good to test that it (the eltype) is `T_prjlvalue` instead of being any `isSpecialPtr`

vtjnash commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+                                    for (unsigned i = 0; i < Phi->getNumIncomingValues(); ++i) {
+                                        BasicBlock *IncombingBB = Phi->getIncomingBlock(i);
+                                        if (DT.dominates(cast<Instruction>(Val), IncombingBB->getTerminator())) {
+                                            Lift->addIncoming(Val, IncombingBB);
```
I'm not confident that LLVM will always store Val using type `T_pprjlvalue`

vtjnash commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1351 @@
+        for (unsigned i = 0; i < Phi->getNumIncomingValues(); ++i) {
+            RecursivelyVisitStoresTo([&](StoreInst *SI){
+                BBState &IncombingBBS = S.BBStates[Phi->getIncomingBlock(i)];
+                NoteUse(S, IncombingBBS, SI->getValueOperand(), IncombingBBS.PhiOuts);
```
The Phi node isn't really a gc-use. I think you can drop this function - it seems you're already walking through the phi nodes when you reach the actual use site (CallInst with jlcall_api in the LocalScan)?

Keno replied:
I think you're right that it's been subsumed by that code. Let me make the change and find out.

Keno committed:
Remove unnecessary MaybeNotePhiJLCallFrameUses

tkelman commented:
I have patched LLVM binaries built, I can upload them and switch appveyor over when I'm a little less occupied with rc3. We may need to do some separate toolchain adjustments to account for opensuse starting to use gcc 7.

Keno committed:
Change IR representation of jlcall frames

Details are in the devdocs. This scheme is signfificantly simpler.

Keno commented:
Upon revisiting this, I was unhappy with the amount of code decicated to working around the fact that even though jlcall operands are ssa uses, they get represented as non-ssa slots. It's like peeling an apple after cutting it. Yeah, you can kinda do it but you can to pick up a million pieces and it's gonna be a mess. I think I came up with a singificantly better way to do it. See devdocs description in the latest commit.

vtjnash commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1232 @@
+    std::vector<int> Colors = ColorRoots(S);
+    std::map<Value *, std::pair<int, int>> CallFrames; // = OptimizeCallFrames(S, Ordering);
+    PlaceRootsAndUpdateCalls(F, Colors, S, CallFrames);
+    assert(!llvm::verifyFunction(F, &dbgs()));
```
drop me (before merging)

vtjnash commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1232 @@
+                    MaybeNoteDef(S, BBS, Phi, BBS.Safepoints);
+                    for (unsigned i = 0; i < Phi->getNumIncomingValues(); ++i) {
+                        BBState &IncombingBBS = S.BBStates[Phi->getIncomingBlock(i)];
+                        NoteUse(S, IncombingBBS, Phi->getIncomingValue(i), IncombingBBS.PhiOuts);
```
IncombingBB

tkelman replied:
oh yeah still an extra b in there

vtjnash commented:
The windows CI builds looks like you have some LLVM assertions that need to be addressed. Beyond that, is this ready to go? Since we run this pass early with system LLVM to avoid its bugs, what's left that's holding this up?

Keno commented:
Nothing left other than @tkelman putting new windows LLVM binaries in place to fix those assertions.

timholy commented:
Would be good to run nanosoldier before merging.

Keno commented:
While we're waiting for new LLVM binaries, there should be no problem with running nanosolider:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-307472399) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/cd0e9ba57f5fcbafc4d4cad989619f5910ad7826/60d0ad2_vs_f541d29/report.md). cc @jrevels

Keno commented:
I'll look into some of the more egregious regression here, but I don't plan to look at everything, because running this pass this early in the pipeline is only a stopgap until we upgrade to a newer LLVM.

Keno commented:
Also, I should probably rebase this and re-run nanosoldier, since it does pick up everything that improved on master since.

Keno commented:
Rebased, so let's have another go at it:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-307512783) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/797c1cb4ce811da02af26830e2b71ccf5c3b6292/3bded76_vs_70915de/report.md). cc @jrevels

Keno committed:
Add a pass to propagate addrspace information

Some LLVM passes don't handle addrspacecast too well, so try to
minimize addrspace cast transitions where legal according to
our invariants.

Keno commented:
Let's try this one. @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-307842053):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/02278420ba54c243c10ef1a27f10f17cb34cdcfd/9e47184_vs_85ef52c/report.md)
cc @jrevels

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-307855329):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/638916db7a61e70394bd1eae313c79bb8072835b/d76a4ba_vs_85ef52c/report.md)
cc @jrevels

Keno commented:
Ok, rebased. Issues that crept up on master have been fixed. Appveyor LLVM is upgraded and nanosoldier's LLVM problem should have been fixed. I think we're getting close here.
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

Keno commented:
I'm seeing some pretty unfortunate compile time regressions on this branch (2x). However, essentially none of that is in the new passes introduced here. For some reason this makes the rest of LLVM a bunch slower. I'm taking a closer look.

Keno commented:
The worst of is is during `-O3` though. My working theory is that the representation in this PR is easier for LLVM to analyze, so it's doing more alias queries, causing the observed performance regression (since we only add basicaa for O3).

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1236 @@
+    else if (getValueAddrSpace(FalseBase) != AddressSpace::Tracked)
+        FalseBase = ConstantPointerNull::get(cast<PointerType>(TrueBase->getType()));
+    Value *SelectBase = SelectInst::Create(SI->getCondition(),
+      TrueBase, FalseBase, "gclift", SI);
```
4 space

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1236 @@
+
+protected:
+    void getAnalysisUsage(AnalysisUsage &AU) const override {
+      FunctionPass::getAnalysisUsage(AU);
```
4 space

tkelman commented on src/intrinsics.cpp:
```diff
@@ -682,7 +682,15 @@ static jl_cgval_t emit_pointerset(jl_cgval_t *argv, jl_codectx_t *ctx)
+                builder.CreateGEP(thePtr, im1), align_nb);
+            tbaa_decorate(tbaa_data, store);
+        } else {
+          typed_store(thePtr, im1, x, ety, ctx, tbaa_data, NULL, align_nb);
```
4 space

tkelman commented on src/codegen.cpp:
```diff
@@ -4061,7 +4061,8 @@ static Value *emit_condition(const jl_cgval_t &condV, const std::string &msg,
     if (condV.isboxed) {
-        return builder.CreateICmpEQ(boxed(condV, ctx), literal_pointer_val(jl_false));
+        return builder.CreateICmpEQ(boxed(condV, ctx),
+          maybe_decay_untracked(literal_pointer_val(jl_false)));
```
4 space

tkelman commented on src/codegen.cpp:
```diff
@@ -3725,9 +3715,11 @@ static jl_cgval_t emit_local(jl_value_t *slotload, jl_codectx_t *ctx)
+            if (v.V) { // v.V will be null if it is a union of all ghost values
+                boxed = decay_derived(boxed);
+                v.V = builder.CreateSelect(load_unbox, emit_bitcast(
+                  decay_derived(v.V), boxed->getType()), boxed);
```
4 space

tkelman commented on src/codegen.cpp:
```diff
@@ -2583,11 +2569,13 @@ static Value *emit_f_is(const jl_cgval_t &arg1, const jl_cgval_t &arg2, jl_codec
         // even if the other pointer managed to get garbage collected
-        return builder.CreateICmpEQ(boxed(arg1, ctx, false), boxed(arg2, ctx, false));
+        return builder.CreateICmpEQ(
+          mark_callee_rooted(boxed(arg1, ctx, false)),
```
4 space

tkelman commented on src/cgutils.cpp:
```diff
@@ -2241,8 +2331,9 @@ static void emit_setfield(jl_datatype_t *sty, const jl_cgval_t &strct, size_t id
-            tbaa_decorate(strct.tbaa, builder.CreateStore(r, emit_bitcast(addr, T_ppjlvalue)));
+            Value *r = maybe_decay_untracked(boxed(rhs, ctx, false)); // don't need a temporary gcroot since it'll be rooted by strct (but should ensure strct is rooted via mark_gc_use)
+            tbaa_decorate(strct.tbaa, builder.CreateStore(r,
+              emit_bitcast(addr, T_pprjlvalue)));
```
4 space

tkelman commented on src/cgutils.cpp:
```diff
@@ -1611,26 +1691,30 @@ static Value *emit_arraylen(const jl_cgval_t &tinfo, jl_value_t *ex, jl_codectx_
     MDNode *tbaa = arraytype_constshape(tinfo.typ) ? tbaa_const : tbaa_arrayptr;
+    if (isboxed) {
+        addr = builder.CreateBitCast(addr,
+          PointerType::get(T_pprjlvalue, cast<PointerType>(addr->getType())->getAddressSpace()));
```
4 space

tkelman commented on src/cgutils.cpp:
```diff
@@ -1363,7 +1438,8 @@ static bool emit_getfield_unknownidx(jl_cgval_t *ret, const jl_cgval_t &strct,
                     builder.CreateLoad(
-                        builder.CreateGEP(data_pointer(strct, ctx), idx)),
+                        builder.CreateBitCast(builder.CreateGEP(decay_derived(data_pointer(strct, ctx)), idx),
+                          PointerType::get(T_prjlvalue, AddressSpace::Derived))),
```
4 space

tkelman commented on src/cgutils.cpp:
```diff
@@ -1049,15 +1113,20 @@ static std::pair<Value*, bool> emit_isa(const jl_cgval_t &x, jl_value_t *type, c
         }
-        return std::make_pair(builder.CreateICmpEQ(emit_typeof_boxed(x, ctx), literal_pointer_val(type)), false);
+        return std::make_pair(builder.CreateICmpEQ(emit_typeof_boxed(x, ctx),
+          maybe_decay_untracked(literal_pointer_val(type))), false);
```
4 space

tkelman commented on src/cgutils.cpp:
```diff
@@ -695,22 +752,27 @@ static jl_cgval_t emit_typeof(const jl_cgval_t &p, jl_codectx_t *ctx)
                     Value *cmp = builder.CreateICmpEQ(tindex, ConstantInt::get(T_int8, idx));
-                    pdatatype = builder.CreateSelect(cmp, literal_pointer_val_slot((jl_value_t*)jt), pdatatype);
+                    pdatatype = builder.CreateSelect(cmp,
+                      decay_derived(emit_bitcast(literal_pointer_val_slot((jl_value_t*)jt), T_ppjlvalue)),
```
4 space

tkelman commented on src/cgutils.cpp:
```diff
@@ -25,6 +25,40 @@ static Value *prepare_call(Value *Callee)
+{
+    assert(V->getType() == T_pjlvalue || V->getType() == T_prjlvalue);
+    return builder.CreateAddrSpaceCast(V,
+      PointerType::get(T_jlvalue, AddressSpace::CalleeRooted));
```
4 space

tkelman commented on src/ccall.cpp:
```diff
@@ -1986,6 +2003,13 @@ jl_cgval_t function_sig_t::emit_a_ccall(
             v = julia_to_address(largty, jargty_in_env, unionall_env, arg,
                                  ai + 1, ctx, &needStackRestore);
+            if (isa<UndefValue>(v)) {
+               JL_GC_POP();
```
4 space

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1236 @@
+                Instruction *StartOfF = &*(F.getEntryBlock().begin());
+                Value *Frame = new AllocaInst(AT,
+#if JL_LLVM_VERSION >= 50000
+                  0,
```
4 space

tkelman commented:
How far out is LLVM 5.0 scheduled to be?

Keno committed:
Better way to expose julia meta-pass

In the previous iteration of this code, timing would be double counted.
This way, the individual passes are correctly registered with the
top level manager, so timing works properly. It's a bit hacky but
with the legacy pass manager, there's not much else to be done.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-308244889) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/9aebfba8943d966f2fb0f1dfd4b0f46aa5c4861b/13732d1_vs_95bf00d/report.md). cc @jrevels

Keno commented:
> How far out is LLVM 5.0 scheduled to be?

If they stick to their regular schedule, branching will be mid July, with final release around September.

tkelman commented:
That's not too bad. Should we do some test nanosoldier runs against llvm trunk to see how much better that'll make things?

Keno commented:
Something is weird with these nanosoldier results. They still include some of the regressions that I explicitly verified as being fixed. Must have broken something. Will take another look.

Keno commented:
Ok, found the bug. I do plan to merge this if nanosoldier results are reasonable, since this PR is getting a bit unwieldy to keep rebasing. The `-O3` compile time perf regression is of course something to be looked at, but I'll do that as a follow up. We might be able to provide our own AA to be a bit faster here. 
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-308288523) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f5048c1407507a0fa5d1b3d9ddb91dff87b90d3e/e77a9e8_vs_493311b/report.md). cc @jrevels

tkelman commented:
Isn't compiling on 32 bit linux

timholy commented:
To me those regressions look like something we could live with temporarily, presuming that we'll work more on the performance once we switch to LLVM5. And I note a good number of performance improvements.

Keno commented:
@tkelman I think travis cached a broken LLVM version there. @timholy I have a fix for most of the remaining regression. I think that's the last thing I'll do on this branch before the merge though.

Keno committed:
Don't color roots that don't need it

Also avoid numbering arguments early in the pipeline. Improves
performance on small test cases without safepoints.

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21888#issuecomment-308477617) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a48d0605944f0388959f588128138a9f2c4651c1/095e9cc_vs_516617d/report.md). cc @jrevels

Keno commented:
x86_64 encountered a network error, which should be unrelated to this PR. OS X looks like it decided to build gcc, which I believe is also unrelated to this PR. @staticfloat why would that happen. I plan to merge this in a couple hours. I'll continue addressing the last couple remaining performance regressions after this is merged.

vtjnash commented:
That's starting to look much better. Still a couple left to fix, though. I know this is a pain to rebase, but please don't merge too quickly. There's many new commits (and llvm passes) on here since it's been last reviewed.

vtjnash commented on src/llvm-propagate-addrspaces.cpp:
```diff
@@ -0,0 +1,233 @@
+        CI->setMetadata(MD.first, MD.second);
+    CI->setDebugLoc(MI.getDebugLoc());
+#endif
+ToInsert.push_back(std::make_pair(CI, &MI));
```
indentation

vtjnash commented on src/llvm-propagate-addrspaces.cpp:
```diff
@@ -0,0 +1,233 @@
+        for (const auto &MD : TheMDs)
+            CI->setMetadata(MD.first, MD.second);
+        CI->setDebugLoc(MTI.getDebugLoc());
+    #endif
```
indentation

vtjnash commented on src/llvm-propagate-addrspaces.cpp:
```diff
@@ -0,0 +1,233 @@
+    Value *TheFn = Intrinsic::getDeclaration(MTI.getModule(), MTI.getIntrinsicID(),
+        {Dest->getType(), Src->getType(),
+         MTI.getOperand(2)->getType()});
+    CallInst *CI = CallInst::Create(TheFn, {Dest, Src,
```
I think it'll be better to just do `MTI->setCalledFunction(TheFn); setArgOperand(0, Dest); setArgOperand(1, Src);`. That way, there's fewer things that need to be moved over to get wrong / outdated.

http://llvm.org/docs/doxygen/html/classllvm_1_1CallInst.html#a61f747edd1ca001427e02e02a320b709

vtjnash commented on src/llvm-propagate-addrspaces.cpp:
```diff
@@ -0,0 +1,233 @@
+        for (;it != GEP->op_end(); ++it) {
+            Operands.push_back(*it);
+        }
+        NewGEP->mutateType(GetElementPtrInst::getGEPReturnType(GEPTy, CurrentV, Operands));
```
I think this clone mutation may be clearer / simpler if expressed as `NewGEP->mutateType(PointerType::getUnqual(cast<PointerType>(GEP->getType())->getElementType(), 0))`

Keno commented:
I've squashed some of the cleanup commits in the middle, addressed the latest round of feedback and addressed the latest rounds of comments. There shouldn't really be any changes to the generated code, but I'll let CI run through just in case (also in case there's interactions with master). I do plan to merge this as soon as CI is through though.

vtjnash commented on src/llvm-propagate-addrspaces.cpp:
```diff
@@ -0,0 +1,205 @@
+        GetElementPtrInst *NewGEP = cast<GetElementPtrInst>(GEP->clone());
+        ToInsert.push_back(std::make_pair(NewGEP, GEP));
+        Type *GEPTy = GEP->getSourceElementType();
+        Type *NewRetTy = cast<PointerType>(GEP->getType())->getElementType()->getPointerTo(getValueAddrSpace(CurrentV));
```
If you're putting an addrspace on this, there should be one on the BitCast below too, otherwise the BitCast inst would be attempting to change AddrSpace also.

Keno replied:
In practice the adddrspace is always 0 (on non-gpu targets anyway), so it's fine, but sure, I'll add the addrspace to the bitcast as well.

Keno replied:
After merging that is. Otherwise, by the time we're through the CI queue again I'll have to rebase yet again.

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+    }
+}
+
+/* For chordal interference graphs, this class gives the verticies in a (reverse
```
vertices

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+                S.LiveSets[idx][Live] = 1;
+        }
+    }
+    // Compute the inference graph
```
inference or interference?

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+}
+
+void LateLowerGCFrame::NoteUse(State &S, BBState &BBS, Value *V, BitVector &Uses) {
+    // Short circuit to have to avoid dealing specially with vectors of
```
Short circuit to avoid having to deal specially with ?

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+    Type *RT = Def->getType();
+    if (isSpecialPtr(RT)) {
+        assert(getValueAddrSpace(Def) == AddressSpace::Tracked &&
+          "Returned value of GC interest, but not tracked?");
```
4 space

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+        S.AllVectorNumbering[V] = std::vector<int>{};
+    }
+    /* We (the frontend) don't insert either of these, but it would be legal -
+       though a bit strange, considering they're pointers) for the optimizer to
```
close paren has no matching open paren

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+    // The result of the local analysis
+    std::map<BasicBlock *, BBState> BBStates;
+
+    // The assignment of numbers to safepoints. The indicies in the map
```
indices

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+    // The reverse of the previous maps
+    std::map<int, Value *> ReversePtrNumbering;
+    // Neighbors in the coloring interferece graph. I.e. for each value, the
+    // indicies of other values that are used simulataneously at some safe point.
```
indices ... simultaneously

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+    std::map<Value *, int> PtrNumbering;
+    // The reverse of the previous maps
+    std::map<int, Value *> ReversePtrNumbering;
+    // Neighbors in the coloring interferece graph. I.e. for each value, the
```
interference

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+    // These do not get updated after local analysis
+    BitVector Defs;
+    BitVector PhiOuts;
+    //// Upward exposed uses that do not have a preceeding safepoint
```
preceding

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+   a path after this point that contains a use of this pointer) is in some gc slot.
+
+   In particular, in order to understand this algorithm, it is important to
+   realize that the only places where rootedness matters is at a safepoint.
```
only place where ... at a safepoint

or

only places where ... at safepoints

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+      and does not look into a basic block. The algorithm is essentially
+      textbook iterative data flow for liveness computation. However, the
+      data flow equations are slightly more complicated because we also
+      forward propagate rootedness information in addition to backproagating
```
backpropagating

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+              set.
+
+       Lastly, we also explicitly compute, for each value, the list of values
+       that are simulataneously live at some safepoint. This is known as an
```
simultaneously

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+      and bloating the size of the stack. Assigning values to these slots is,
+      equivalent to doing graph coloring on the interference graph - the graph
+      where nodes are values and two values have an edge if they are
+      simulataneously live at a safepoint - which we computed in the previous
```
simultaneously

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+
+    4. GC Root coloring
+
+      Two values which are not simulataneously live at a safepoint can share the
```
simultaneously

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+      Two values which are not simulataneously live at a safepoint can share the
+      same slot. This is an important optimization, because otherwise long
+      functions would have exceptionally large GC slots, reducing performance
+      and bloating the size of the stack. Assigning values to these slots is,
```
no comma after is

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+      (this is beneficial for code where the primary path does not have
+      safepoints, but some other path - e.g. the error path does). However,
+      if the first safepoint is not dominated by the definition (this can
+      happen due to the non-ssa corner cases), the store is insert right after
```
is inserted

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+      of the algorithm and their operands as uses of those values. It is
+      important to consider however WHERE the uses of PHI's operands are
+      located. It is neither at the start of the basic block, because the values
+      do not dominated the block (so can't really consider them live-in), nor
```
do not dominate

tkelman commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -0,0 +1,1250 @@
+                        %Bbase, %B]
+
+      We then pretend, for the purposes of numbering that %phi was derived from
+      %philift. Note that in order to be able to this, we need to be able to
```
in order to be able to do this,

tkelman reviewed (commented):
few minor typos etc, can be addressed in either a ci skip commit or later followup

Keno commented:
Will do various cleanups in a followup PR.
